man/RNApaln.1.in
.\" .ER
.TH "RNAPDIST" "l" "" "Ivo Hofacker" "Vienna RNA"
.SH "NAME"
RNApaln \- RNA alignment based on sequence base pairing propensities
.SH "SYNOPSIS"
\fBRNApaln [\-Xp|m|f|c] [\-B\ [file]] [--gapo open] [--gape ext] [--seqw w] [--endgaps] [\-T\ temp] [\-4] [\-d] [\-noGU] [\-noCloseGU] [\-e\ e_set] [\-P\ paramfile] [\-nsp\ pairs]

.SH "DESCRIPTION"
.I RNApaln
Uses string-alignment techniques to perform fast pairwise structural alignments of RNAs. Similar to RNApdist secondary structure is incorporated in an approximate manner by computing base pair probabilities, which are then reduced to a vector holding the probability that a base is paired upstream, downstream, or remains unpaired. Such pair propsensity vectors can then be compared using standard alignment algorithms. In contrast to RNApdist, RNApaln performs similarity (instead of distance) alignments, considers both sequence and structure information, and uses affine (rather than linear) gap costs. RNApaln can perform semi-local alignments by using free end gaps, a true local alignment mode is planned.

The same approach has since been used in the StraL program from Gerhard Steeger's group. Since StraL has optimized parameters and a multiple alignment mode, it be be currently the better option. 

.SH "OPTIONS"
.IP \fB\-Xp\fP
compare the structures pairwise (p), that is first with 2nd, third
with 4th etc. This is the default.
.IP \fB\-Xm\fP
calculate the distance matrix between all structures. The output is
formatted as a lower triangle matrix.
.IP \fB\-Xf\fP
compare each structure to the first one.
.IP \fB\-Xc\fP
compare continuously, that is i\-th with (i+1)th structure.
.IP \fB\-B\ [\fI<file>\fP]
Print an "alignment" with gaps of the profiles, using the following symbols:
.br 
(  )  essentially upstream (downstream) paired bases
.br 
{  }  weakly upstream (downstream) paired bases
.br 
 |    strongly paired bases without preference
.br 
 ,    weakly paired bases without preference
.br 
 .    essentially unpaired bases.
.br 

The aligned structures are written to \fI<file>\fP, if specified. Otherwise
output is written to stdout, unless the \-Xm option is set in which case
"backtrack.file" is used.
.TP
.B \--gapo \fIopen\fP
Set the gap open penalty to \fIopen\fP
.B \--gape \fIext\fP
Set the gap extension penalty to \fIext\fP
.B \--seqw \fIw\fP
Set the weight of sequence (compared to structure) in the scoring function.
.B \--endgaps
Use free end-gaps.

.PP 
The \fB\-T\fP, \fB\-d\fP, \fB\-4\fP, \fB\-noGU\fP, \fB\-noCloseGU\fP,
\fB\-e\fP, \fB\-P\fP, \fB\-nsp\fP, options work as in RNAfold.

.SH "REFERENCES"
Bonhoeffer S, McCaskill J S, Stadler P F, Schuster P, (1993)
RNA multi\-structure landscapes, Euro Biophys J:22,13\-24
.SH "VERSION"
This man page documents version @VERSION@ Vienna RNA Package.
.SH "AUTHORS"
Peter F Stadler, Ivo L Hofacker, Sebastian Bonhoeffer.
.SH "BUGS"
We hoped you wouldn't find them.
.br 
Comments should be sent to rna@tbi.univie.ac.at




NEWS
Upcoming 
-----------
- Updated parameters for both RNA and DNA using the data compiled by 
  David Mathews <David_Mathews@urmc.rochester.edu>. 

Version 1.6
------------
- The RNAforester program for tree-alignments of RNA structures is now
  distributed with the Vienna RNA package, see the RNAforester
  subdirectory for more information. RNAforester was written by Matthias
  Hoechsmann <mhoechsm@techfak.uni-bielefeld.de> 
- The Kinfold program for stochastic simulation of folding trajectories is
  no included in the package, see the Kinfold subdirectory.
- cofolding of two structures now supports suboptimal folding and 
  partition function folding (still experimental). 
  ATTENTION: Energies of hybrid structures now inlcude the Duplex-initiation
  energy, which was neglected in previous version.
- RNAplfold is a partition function variant of RNALfold. It computes the
  mean probability of a (local) base pair averaged over all sequence
  windows that contain the pair.
- new utilities to color alignments and consensus structures  

Version 1.5pre 
-----------
- ViennaRNA now uses autoconfig generated configure scripts for even better
  portability (should compile on any UNIX, Linux, MacOS X, Windows with
  Cygwin).
- The new RNAalifold program predicts consensus structures for a set of
  aligned sequences.
- Complete suboptimal folding is now integrated in the library.
- Beginning support for co-folding of two strands: energy_of_struct() and
  RNAeval can now compute energies of duplex structures.
- RNAcofold predicts hybrid structures of two RNA strands
- RNAduplex predicts hybrid structures, while allowing only inter-molecular
  base pairs (useful for finding potential binding sites)
- RNALfold predicts locally stable structures in long sequences.
- Major changes to Perl module. See the pod documentation (perldoc RNA).
- RNAsubopt can do stochastic backtracking to produce samples of suboptimal
  structures with Boltzmann statistics.
- New utilities to rotate secondary structure plots and annotate them with
  reliability data.
- Various small bug fixes

Version 1.4
-----------
- New Turner parameters as described in Mathews et.al. JMB v288, 1999.
  Small changes to format of parameter files (old param files won't work!)
- mfe and suboptimal folding will produce only structures without isolated
  pairs if noLonelyPairs=1 (-noLP option), for partition function folding
  pairs that can only occur as isolated pairs are not formed.
- setting dangles=3 (-d3 option) will allow co-axial stacking of adjacent
  helices in mfe folding and energy_of_struct().

Version 1.3.1
-------------
- RNAheat would produce spikes in the specific heat because dangling end
  energies did not go smoothly to 0. 
- PS dot plots now have an option to use a log scale (edit _dp.ps file
  and set logscale to true).

Version 1.3
-----------
- Secondary structure plots now use E. Bruccoleri's naview routines for
  layout by default. New utility RNAplot produces secondary structure plots 
  from structures in bracket notation with several options.
- New -d2 option in RNAfold and RNAeval sets dangles=2, which makes 
  energy_of_struct() and fold() treat dangling ends as in pf_fold().
  -noLP option in RNAfold etc sets noLonelyPairs=1, which avoids most
  structures containing lonely base pairs (helices of length 1).
- new utility functions pack_structure() unpack_structure() make_pair_table()
  and bp_distance().  RNAdistance adds bp_distance() via -DP switch.
- First release of RNAsubopt for complete suboptimal folding.
- fixed bug in asymmetry penalty for interior loops. 
- Default compilation now uses doubles for partition function folding.

Version 1.2.1
-------------
- Fixed bug in version 1.2 of the RNAheat program causing overflow errors
  for most input sequences. 
- The PS_dot_plot() and PS_rna_plot() routines now return an int. The return 
  value is 0 if the file could not be written, 1 otherwise.
- This version contains the alpha version of a perl5 module, which let's you
  access all the capabilities of the Vienna RNA library from perl scripts.

Version 1.2
-----------
- New energy parameters from (Walter et.al 1994). 
- Energy parameters can be read from file.
- RNAeval and energy_of_struct() support logarithmic energy function for 
  multi-loops.
- gmlRNA() produces secondary structure drawing in gml (Graph Meta Language). 
- Many bug fixes.


Progs/RNApaln.c
/*
		Distances of Secondary Structure Ensembles
	  Peter F Stadler, Ivo L Hofacker, Sebastian Bonhoeffer
			Vienna RNA Package
*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include "part_func.h"
#include "fold.h"
#include "fold_vars.h"
#include "profiledist.h"
#include "dist_vars.h"
#include "utils.h"
#include "ProfileAln.h"

#define PUBLIC
#define PRIVATE    static

#define MAXLENGTH  10000
#define MAXSEQ      1000
/*@unused@*/
static char rcsid[] = "$Id: RNApaln.c,v 1.3 2005/07/24 08:35:15 ivo Exp $";

static  double gapo=1.5, gape=0.666, seqw=0.5;
static  int endgaps=0;  

PRIVATE void command_line(int argc, char *argv[]);
PRIVATE void usage(void);
PRIVATE void print_aligned_lines(FILE *somewhere);

PRIVATE char task;
PRIVATE char outfile[50];
PRIVATE char  ruler[] ="....,....1....,....2....,....3....,....4"
                       "....,....5....,....6....,....7....,....8";

extern void  PS_dot_plot(char *string, char *file);
extern void  read_parameter_file(const char fname[]);
extern float profile_aln(const float *T1, const char *seq1, 
			 const float *T2, const char *seq2);
static int noconv = 0;

int main(int argc, char *argv[])
     
{
  float     *T[MAXSEQ];
  char      *seq[MAXSEQ];
  int        i,j, istty, n=0;
  int        type, length, taxa_list=0;
  float      dist;
  FILE      *somewhere=NULL;
  char      *structure;
  char      *line=NULL, fname[20], *list_title=NULL;

  command_line(argc, argv);
   
  if((outfile[0]=='\0')&&(task=='m')&&(edit_backtrack)) 
    strcpy(outfile,"backtrack.file"); 
  if (outfile[0]!='\0') somewhere = fopen(outfile,"w");
  if (somewhere==NULL) somewhere = stdout;
  istty   = (isatty(fileno(stdout))&&isatty(fileno(stdin)));

  while (1) {
    if ((istty)&&(n==0)) {
      printf("\nInput sequence;  @ to quit\n");
      printf("%s\n", ruler);
    }

    type = 0;
    do {  /* get sequence to fold */
      if (line!=NULL) free(line);
      *fname='\0';
      if ((line=get_line(stdin))==NULL) {type = 999; break;}
      if (line[0]=='@') type = 999;
      if (line[0]=='*') {
	if (taxa_list==0) {
	  if (task=='m') taxa_list=1;
	  printf("%s\n", line);
	  type = 0;
	} else {
	  list_title = strdup(line);
	  type = 888;
	}
      }
      if (line[0]=='>') {
	if (sscanf(line,">%12s", fname)!=0)
	  strcat(fname, "_dp.ps");
	if (taxa_list)
	  printf("%d : %s\n", n+1, line+1);
	else printf("%s\n",line);
	type = 0;
      }
      if (isalpha(line[0]))  {
	char *cp;
	cp =strchr(line,' ');
	if (cp) *cp='\0';
	type = 1;
      }
    } while(type==0);

    if( (task == 'm')&&(type>800) ) {
      if (taxa_list) 
	printf("* END of taxa list\n");
      printf("> p %d (pdist)\n",n);
      for (i=1; i<n; i++) {
	for (j=0; j<i; j++) {
	  printf("%g ",profile_aln(T[i],seq[i], T[j],seq[j]));
	  if(edit_backtrack) fprintf(somewhere,"> %d %d\n",i+1,j+1);
	  print_aligned_lines(somewhere);
	}
	printf("\n");
      }
      if (type==888) {  /* do another distance matrix */
	n = 0;
	printf("%s\n", list_title);
	free(list_title);
      }
    }      
      
    if(type>800) {
      for (i=0; i<n; i++) 
	free_profile(T[i]);
      if (type == 888) continue;
      if (outfile[0]!='\0') (void) fclose(somewhere);
      if (line!= NULL) free(line);
      return 0; /* finito */
    }
      
    length = (int) strlen(line);
    for (i=0; i<length; i++) {
      line[i]=toupper(line[i]);
      if (!noconv && line[i] == 'T') line[i] = 'U';
    }

    {
      double mfe, kT;
      kT = (temperature+273.15)*1.98717/1000.; /* in Kcal */
      structure = (char *) space((length+1)*sizeof(char));
      mfe = fold(line, structure);
      pf_scale = exp(-(1.07*mfe)/kT/length);
      init_pf_fold(length);
      (void) pf_fold(line,structure);
    }

    if (*fname=='\0')
      sprintf(fname, "%d_dp.ps", n+1);
    PS_dot_plot(line, fname);

    T[n] = Make_bp_profile(length);
    seq[n] = strdup(line);
    if((istty)&&(task=='m')) printf("%s\n",structure);
    free(structure);
    
    free_arrays();
    free_pf_arrays();
      
    n++;
    switch (task) {
    case 'p' :
      if (n==2) {
	dist = profile_aln(T[0],seq[0],T[1],seq[1]);
	printf("%g\n",dist);
	print_aligned_lines(somewhere);
	free_profile(T[0]);
	free_profile(T[1]);
	free(seq[0]); free(seq[1]);
	n=0;
      }
      break;
    case 'f' :
      if (n>1) { 
	dist = profile_aln(T[1], seq[1], T[0], seq[0]);
	printf("%g\n",dist);
	print_aligned_lines(somewhere);
	free_profile(T[1]); free(seq[1]);
	n=1;
      }
      break;
    case 'c' :
      if (n>1) {
	dist = profile_aln(T[1], seq[1], T[0],seq[0]);
	printf("%g\n",dist);
	print_aligned_lines(somewhere);
	free_profile(T[0]); free(seq[0]);
	T[0] = T[1]; seq[0] = seq[1];
	n=1; 
      }
      break;
	  
    case 'm' : 
      break;
	  
    default :
      nrerror("This can't happen.");
    }    /* END switch task */
    (void) fflush(stdout);
  }    /* END while */
  if (line !=NULL) free(line);
  return 0;
}

/* ----------------------------------------------------------------- */

PRIVATE void command_line(int argc, char *argv[])
{

  int i, sym;
  char *ns_bases=NULL, *c;
  char *ParamFile=NULL;

  task = 'p';
  for (i=1; i<argc; i++) {
    if (argv[i][0]=='-') {
      switch (argv[i][1]) {
      case 'T':      /* temperature for folding */
	if (argv[i][2]!='\0') usage();
	if (sscanf(argv[++i], "%lf", &temperature)==0)
	  usage();
	break;
      case '4':
	tetra_loop=0;
	break;
      case 'd':
	dangles=0;
	break;
      case 'e':
	if (sscanf(argv[++i],"%d", &energy_set)==0)
	  usage();
	break;
      case 'n':
	if ( strcmp(argv[i], "-noGU" )==0) noGU=1;
	if ( strcmp(argv[i], "-noCloseGU" ) ==0) no_closingGU=1;
	if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	if ( strcmp(argv[i], "-nsp") ==0) {
	  if (++i<argc)
	    ns_bases = argv[i];
	  else  usage();
	}
	if ( strcmp(argv[i], "-noconv")==0) noconv=1;
	break;
      case 'X':
	switch (task = argv[i][2]) {
	case 'p': break;
	case 'm': break;
	case 'f': break;
	case 'c': break;
	default : usage();
	}
	break;
      case 'B':
	if(argv[i][2]!='\0') usage();
	if( (i+1) >= argc) outfile[0] = '\0';
	else if (argv[i+1][0]=='-') outfile[0] = '\0';
	else {
	  i++;
	  strncpy(outfile,argv[i],49);
	}
	edit_backtrack = 1;   
	break;
      case 'P':
	if (++i<argc)
	  ParamFile=argv[i];
	else usage();
	break;
      case '-':
	if (strcmp(argv[i], "--gapo")==0) {
	  if (sscanf(argv[++i],"%lf", &gapo)==0)
	    usage();
	} else {
	  if (strcmp(argv[i], "--gape")==0) {
	    if (sscanf(argv[++i],"%lf", &gape)==0)
	      usage();
	  } else {
	    if (strcmp(argv[i], "--seqw")==0) {
	      if (sscanf(argv[++i],"%lf", &seqw)==0)
		usage();
	    } else {
	      if (strcmp(argv[i], "--endgaps")==0) 
		endgaps=1;
	    }
	  }
	}
	break;
	
      default:
	usage();
      }
    }
  }
  /* fprintf(stderr, "%f %f %f %d\n", gapo, gape, seqw, -endgaps); */
  set_paln_params(gapo, gape, seqw, 1-endgaps);

  if (ParamFile!=NULL)
    read_parameter_file(ParamFile);

  if (ns_bases!=NULL) {
    nonstandards = space(33);
    c=ns_bases;
    i=sym=0;
    if (*c=='-') {
      sym=1; c++;
    }
    while (*c!='\0') {
      if (*c!=',') {
	nonstandards[i++]=*c++;
	nonstandards[i++]=*c;
	if ((sym)&&(*c!=*(c-1))) {
	  nonstandards[i++]=*c;
	  nonstandards[i++]=*(c-1);
	}
      }
      c++;
    }
  }
}

/* ------------------------------------------------------------------------- */

PRIVATE void usage(void)
{
  nrerror
    ("usage: RNApaln [-Xpmfc] [-B [file]] [-T temp] [-4] [-d] [-noGU]\n"
     "               [-noCloseGU] [-noLP] [-P paramfile] [-nsp pairs]\n"
     "               [--gapo open] [--gape ext] [--seqw w] [--endgaps]\n");
}

/*--------------------------------------------------------------------------*/

PRIVATE void print_aligned_lines(FILE *somewhere)
{
  if (edit_backtrack)
    fprintf(somewhere, "%s\n%s\n%s\n%s\n",
	    aligned_line[2], aligned_line[0],
	    aligned_line[3], aligned_line[1]);
}

/*--------------------------------------------------------------------------*/
